_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+is the basis of [multi-platform software](https://en.wikipedia.org/wiki/Cross-platform_software).
+
+*Toolchains* are the set of programs that build your code. A C++ toolchain
+includes a compiler and linker and the flags that invoke them "correctly".
+Different platforms require different toolchains. gcc might be the compiler of
+choice for Linux while Xcode is used on the Mac. Even the same platform might
+use different toolchains. For example, maybe you want to use an experimental
+compiler.
+
+Not only are these concepts cross-language, but it's *important* that languages
+treat them consistently. Otherwise you get an uncoordinated mess of
+language-specific code that makes it hard to combine different languages in a
+single project.
+
+One of the reasons
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+doesn't work well is that `--cpu` and `--crosstool_top` are C++-specific. What
+if you want to build Java libraries with C++ dependencies for both Linux and
+Android? This impacts not just C++, but also the JDK and maybe even which
+support libraries you need.
+
+Not only are `--cpu` and `-crosstool_top` insufficiently expressive, but they
+might actively sabotoge you. Consider this use of
+[select](https://docs.bazel.build/versions/master/configurable-attributes.html):
+
+```python
+java_library(
+    name = "supporting_library",
+	deps = select({
+	    ":android": [":extra_required_android_deps"],
+		"//conditions:default": []
+    }))
+```
+
+If you use `--cpu` and `crosstool_top` to define what `:android` means, what
+happens when an app supports a new Android phone with a different CPU?
+`:android` won't trigger, the app won't get required Android support libraries,
+and it will crash.
+
+### A Better Way
+
+Configurability's platform work, led by [@katre](https://github.com/katre),
+lets you declare exactly what you want in a way all rules understand. It
+doesn't even matter if no one's ever heard of your platform.
+
+With the new API, our example can be rewritten as:
+
+```python
+$ cat platforms/BUILD
+constraint_setting(name = "os")
+constraint_value(name = "android", constraint_setting = ":os")
+constraint_value(name = "linux", constraint_setting = ":os")
+
+constraint_setting(name = "device") constraint_value(name = "phone",
+constraint_setting = ":device") constraint_value(name = "tablet",
+constraint_setting = ":device")
+
+platform(name = "pixel3", constraint_values = [":android", ":phone"])
+```
+
+```python
+$ cat helpers/BUILD
+config_setting(name = "android", constraint_values = "//platforms:android")
+java_library(
+    name = "supporting_library",
+    deps = select({
+	    ":android": [":extra_required_android_deps"],
+        "//conditions:default": []
+    }))
+```
+
+and built with:
+
+    $ bazel build //my:binary --platforms=//platforms:pixel3 --define MYFEATURE=1
+
+This is more concise and correct than before. Now it doesn't matter what the
```

Comment:
This is great, and I think you could bring the point home even more strongly: "The benefits here are significant: You no longer have to care about which CPU is being used, and _any_ platform labeled `:android` is an Android platform.

Reply:
Done.